scripts/image-saver.py

```diff
@@ -1,85 +1,179 @@
+    save them in an hdf5 file at the specified path with the
+    as filename if it not empty; if it is empty, it will be saved
+    {camera.name}-{timestamp}. Based on CamH5.collect from
+    /reg/g/pcds/pyps/apps/hutch-python/rix/rix/cam_to_file.py.
 
     Parameters
     ----------
-    args: argparse.Namespace
-        Gives access to arguments from main - cam, num_images, path, filename,
-        which are of type string, int, string, and string. cam and path are
-        required arguments, num_images and filenames have defaults. If filename
-        is blank (len(filename)==0) h5_img_collect will create the file as
-        camera-timestamp; else, the current timestamp will be appended to the
-        filename when this function is called (every time the QPushButton is
-        pressed), ensuring that each button press creates a uniquely named
-        file.
-    h5_img_collect: Callable Object
-        A callable object equivalent to hutch_module.cam_to_file.h5_img_collect
-        that takes arguments camera, images, path, and file_name (all but
-        camera are keyword). camera is the name of camera, images is the number
-        of images to take, path and file_name specify how to save the hdf5
-        file. If file_name is blank (an empty string), it will default to
-        camera-timestamp.
+    camera: pcdsdevices.areadetector.detectors.PCDSAreaDetector
+        The actual camera object containing the hdf51 plugin used to
+        collect and save images. Created by build_cam in cam_load.
+    num_images: int
+        The number of images to take.
+    path: string
+        Path where the hdf5 file containing the images should be saved to.
+    filename: string
+        Name of hdf5 file images are saved in. If empty, will default to
+        {camera}-{timestamp}. If not empty, timestamp will be appended
+        to ensure all captures have unique filenames.
     """
-    current_name = args.filename
+    # Default filename is camera-timestamp
+    if not filename:
+        filename = f'{camera.name}-{int(time())}'
+    # Setting up hdf51 plugin
+    camera.hdf51.file_name.put(filename)
+    camera.hdf51.file_template.put('%s%s_%03d.h5')
+    camera.hdf51.num_capture.put(num_images)
+    camera.hdf51.auto_save.put(True)
+    camera.hdf51.file_write_mode.put('Capture')
+    camera.hdf51.enable.put('Enabled')
+    camera.hdf51.capture.put('Capture')
+    # Creating callbacks to know when collecting/saving images is complete
+    collecting = True
+    writing = True
+
+    def _collecting_cb(value, **kwargs):
+        """Get a callback for collecting changes"""
+        nonlocal collecting
+        collecting = value
+
+    def _writing_cb(value, **kwargs):
+        """Get a callback when writing changes"""
+        nonlocal writing
+        writing = value
+    # Start collecting and saving
+    collect_uid = camera.hdf51.capture.subscribe(_collecting_cb)
+    write_uid = camera.hdf51.write_file.subscribe(_writing_cb)
```

Comment:
It's unlikely, but the code here could hit a race condition if the camera IOC doesn't behave as you'd expect- for example, if these callbacks get run based on the values from a previous function call.

Reply:
Not sure I understand how a race condition could be caused by a previous function call, since each time the script is run a new camera object is made with its own hdf51 member, and since I am not multithreading there will only ever be one of each of the two subscribers. I (briefly) looked at some of the implementation of EpicsSignalWithRBV in ophyd/signal.py, and I think that if you had multiple subscribers, (in different processes) there could be undefined behavior, but I don't see a clean solution since you would need some way of obtaining pv locks at the channel access level. Almost everything hdf51-related in take_images is from CamH5.collect, which is what h5_img_collect uses, which is why I didn't suspect any issues. I will look into this more, but if you have any more thoughts on what I could do to try and fix this, I'm all ears.

Reply:
I'm not sure if you absolutely need to fix this, since it will be very rare and I can't prove that it will happen. This is the sort of thing where I'm more bringing awareness that "code that looks like this can be suspect to race conditions".
Your script and all the objects inside it are allocated each time it is run, but the camera and its EPICS IOC continue to exist and hold some state information.

Consider the following:
1. You run this script a few times
2. On one of the runs, the IOC lags a bit or the network has a bad time, and the IOC does not reset the `capture` and `write_file` PV values (or, the script doesn't see the reset) prior to these subscriptions being set up, so they are still in the final states from the previous script run.
3. The subscriptions see the stale values and conclude that the collections is already done, when it is not.

Again, this is just for awareness that these sorts of patterns can cause race conditions for scripts that interact with EPICS IOCs.

Reply:
Thanks for the clarification. If I wanted to fix this, or similar situations that would occur when interacting with EPICS IOCs, how would I go about doing that? Would using Archiver data or polling for the hdf5 file's creation be a better way to do this? Also should I be worried about multiple processes simultaenously using the hdf5 plugin for the same camera and interfering with each other?

Reply:
For similar situations I would probably do one of or some combination of the following depending on the situation:
- Start the subscription prior to making any changes so it always starts in the same state
- use the `old_value` callback kwarg in addition to the `value` callback kwarg
- rely on the `timestamp` callback kwarg

I think for this specific situation polling the hdf5 file's creation is a pretty reasonable solution too, that shouldn't ever happen out of sequence (but I guess it could take some moments to propagate through the network file system?)

I think your worry about multiple processes is valid. I'm not entirely sure how to approach that one. In general we'd want to make some multi-process mutex but in this case you're not actually in control of the other processes and you're not guaranteed that this is the only script that will use the hdf5 plugin... I think it's OK to disregard it in favor of putting your efforts elsewhere, but I'd approach it from the angle of trying to check if someone else is using the plugin and giving a reasonable error message.